[MIG][IMP] delivery_multi_destination: Introduce 'based on destination' delivery type

carmenbianca opened the pull request:
Backported from #871 
Internal task T12787

website_sale_delivery checks whether delivery_type is set to 'fixed' for some behaviours. And before this commit, multi-destination carriers effectively always had the delivery_type set to 'fixed'. However, because children of the multi-destination carrier can be something other than 'fixed', this causes some strange behaviours.

(Example, on the delivery selection page, you expect 'Select to compute delivery rate' to show up for non-'fixed' carriers, but this doesn't happen.)

We can't make website_sale_delivery aware of destination_type, but we _can_ introduce a non-'fixed' delivery type. So that's what we do here.

The choice was made to turn the destination type into a computed value dependent on the delivery type because there is strong coupling between the two values, and trying to keep the two values in sync using onchange/constrains/etc resulted in terrible spaghetti code.

remytms committed:
[IMP] delivery_multi_destination: Introduce 'based on destination' delivery type

website_sale_delivery checks whether delivery_type is set to 'fixed' for
some behaviours. And before this commit, multi-destination carriers
effectively always had the delivery_type set to 'fixed'. However,
because children of the multi-destination carrier can be something other
than 'fixed', this causes some strange behaviours.

(Example, on the delivery selection page, you expect 'Select to compute
delivery rate' to show up for non-'fixed' carriers, but this doesn't
happen.)

We can't make website_sale_delivery aware of destination_type, but we
_can_ introduce a non-'fixed' delivery type. So that's what we do here.

The choice was made to turn the destination type into a computed value
dependent on the delivery type because there is strong coupling between
the two values, and trying to keep the two values in sync using
onchange/constrains/etc resulted in terrible spaghetti code.

Co-Authored-By: Carmen Bianca BAKKER <carmen@coopiteasy.be>
Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

polchampion commented:
@carmenbianca tested functionally [on the runboat](http://oca-delivery-carrier-16-0-pr887-88a7398d1f76.runboat.odoo-community.org/shop/payment)

- I created some multiple destination delivery methods with non fixed children
- On the delivery selection page, the available carrier rate is computed right away
- The other carrier rates are displayed loading instead of being hidden
- I get the error below on the delivery selection page:

![screenshot-oca-delivery-carrier-16-0-pr887-88a7398d1f76 runboat odoo-community org-2024 10 29-16_28_06](https://github.com/user-attachments/assets/0c07a819-1f69-4346-aa4e-c0a7fc0d678a)


```UncaughtClientError > TypeError

Uncaught Javascript Error > browser.navigator.clipboard is undefined

TypeError: browser.navigator.clipboard is undefined
    onClickClipboard@http://oca-delivery-carrier-16-0-pr887-88a7398d1f76.runboat.odoo-community.org/web/assets/641-afd6f82/1/web.assets_frontend_lazy.min.js:6760:20
    mainEventHandler@http://oca-delivery-carrier-16-0-pr887-88a7398d1f76.runboat.odoo-community.org/web/assets/641-afd6f82/1/web.assets_frontend_lazy.min.js:2063:77
    listener@http://oca-delivery-carrier-16-0-pr887-88a7398d1f76.runboat.odoo-community.org/web/assets/641-afd6f82/1/web.assets_frontend_lazy.min.js:1233:15```

huguesdk commented:
@polchampion could you try again? it seems that the error you reported is unrelated to the changes. how to reproduce it?

polchampion commented:
@huguesdk Can't reproduce the above error, but I tested anew and couldn't make it work.
Tested on [this test db](https://odoo16.main.test.srv.coopiteasy.be/web/login?db=rotor-v16-test).

- I created a multi-destination delivery method called "xtra multi", with two non-fixed children. The price calculation is based on rules and depends on the product weight.
- Odoo can neither calculate the price [in the backend](https://odoo16.main.test.srv.coopiteasy.be/web?db=rotor-v16-test#id=43&cids=1&menu_id=236&action=386&model=sale.order&view_type=form) nor on the website:

![screenshot-odoo16 main test srv coopiteasy be-2025 02 25-10_50_40](https://github.com/user-attachments/assets/3c2617c9-14b8-4710-be6b-57b0475ddbfe)

![screenshot-odoo16 main test srv coopiteasy be-2025 02 25-10_50_47](https://github.com/user-attachments/assets/23c5dd6c-c65b-4d60-8ed0-0c95a2763591)

 
![screenshot-odoo16 main test srv coopiteasy be-2025 02 25-10_51_44](https://github.com/user-attachments/assets/f8497bde-d543-45c9-bc08-6df6aa9f632e)


I'm perhaps wrong on the multi-destination usage, but I'm otherwise not sure how to best test the wanted behaviour. I believe the test should work in any case.

huguesdk committed:
[IMP] query the context in .name_search()

* allow to control the .name_search() behavior with the
  "show_children_carriers" context option.
* rename the args argument of .search() to domain, as in the parent
  method.

huguesdk committed:
[IMP] define proper delivery.carrier methods

* define .base_on_destination_rate_shipment() and
  .base_on_destination_send_shipping() instead of overriding
  .rate_shipment() and .send_shipping() and filtering on
  destination_type.
* always return a value in .base_on_destination_rate_shipment() to fail
  nicely if no sub-carrier matches.

huguesdk commented:
`delivery_multi_destination` is only about defining different carriers depending on the shipping destination. i think that the rules only allow to compute a different price for the same carrier.

polchampion reviewed (approved):
Functionally tested and LGTM

huguesdk commented:
@OCA/logistics-maintainers could someone review or merge this? i could review it myself, but i made some changes myself, so interest is conflicting here.

hparfr commented on delivery_multi_destination/wizards/choose_delivery_carrier.py:
```diff
@@ -0,0 +1,18 @@
+        if self.delivery_type == "base_on_destination":
+            vals = self._get_shipment_rate()
+            if vals.get("error_message"):
+                return {"error": vals["error_message"]}
```
what  does {"error":} do and where ?

huguesdk replied:
thanks for the review! this reproduces [the behavior of the parent model](https://github.com/OCA/OCB/blob/16.0/addons/delivery/wizard/choose_delivery_carrier.py#L34) (in the `delivery` module), but after some digging, i found that an `onchange` method can return a dictionary with a `warning` key (as explained in [the doc](https://www.odoo.com/documentation/16.0/developer/reference/backend/orm.html#odoo.api.onchange)), but that nothing will happen if the dictionary contains an `error` key (as can be seen in [the code handling the return value](https://github.com/OCA/OCB/blob/16.0/odoo/models.py#L6357)). good catch!

i also noticed that the dictionary returned by `._get_shipment_rate()` [can contain a `warning_message` key](https://github.com/OCA/OCB/blob/16.0/addons/delivery/models/delivery_carrier.py#L180). what would you suggest? something like this?

```python
            if vals.get("error_message"):
                return {
                    "warning": {"title": _("Error"), "message": vals["error_message"]}
                }
            if vals.get("warning_message"):
                return {"warning": {"message": vals["warning_message"]}}
```

hparfr replied:
Yes sounds good.

huguesdk committed:
[FIX] correctly handle error and warning

hparfr commented:
@huguesdk please rebase

huguesdk commented:
@hparfr i started rebasing to make it mergeable, but was interrupted. it’s done now.

hparfr reviewed (commented):
/ocabot merge nobump

huguesdk commented:
@hparfr many thanks!
